### ISSUE TITLE
kmod::setting broken for CentOS 7

### ISSUE THREAD

**Issue by baurmatt** [camptocamp/puppet-kmod#45]
## Expected Behavior
Given the following Puppet Code:

```
      $ipmi_kernel_modules = ['ipmi_devintf', 'ipmi_si', 'ipmi_msghandler']

      $ipmi_kernel_modules.each |String $module| {
        kmod::alias { "acpi:IPI000*:* - ${module}":
          source    => $module,
          aliasname => 'acpi:IPI000*:*',
          file      => '/etc/modprobe.d/ipmitool-modalias.conf',
        }
      }
```

```
[11:53:11] root@foo.example.org ~ # grep alias /etc/modprobe.d/ipmitool-modalias.conf
alias acpi:IPI000*:* ipmi_si
alias acpi:IPI000*:* ipmi_devintf
alias acpi:IPI000*:* ipmi_msghandle
```

## Current Behavior
```
[11:56:54] root@bar.example.org~ # really_run_puppet_agent --dev testbranch -f
Info: Using configured environment 'testbranch'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Loading facts
Info: Caching catalog for bar.examplre.org
Info: Applying configuration version '1528105705'
Notice: Augeas[kmod::setting kmod::alias acpi:IPI000*:* - ipmi_devintf acpi:IPI000*:*](provider=augeas):
--- /etc/modprobe.d/ipmitool-modalias.conf	2018-06-04 11:49:41.411437023 +0200
+++ /etc/modprobe.d/ipmitool-modalias.conf.augnew	2018-06-04 11:57:14.466415896 +0200
@@ -1,2 +1,2 @@

-alias acpi:IPI000*:* ipmi_msghandler
+alias acpi:IPI000*:* ipmi_devintf

Notice: /Stage[main]/Ipmitool::Kmod/Kmod::Alias[acpi:IPI000*:* - ipmi_devintf]/Kmod::Setting[kmod::alias acpi:IPI000*:* - ipmi_devintf]/Augeas[kmod::setting kmod::alias acpi:IPI000*:* - ipmi_devintf acpi:IPI000*:*]/returns: executed successfully
Notice: Augeas[kmod::setting kmod::alias acpi:IPI000*:* - ipmi_si acpi:IPI000*:*](provider=augeas):
--- /etc/modprobe.d/ipmitool-modalias.conf	2018-06-04 11:57:14.490415736 +0200
+++ /etc/modprobe.d/ipmitool-modalias.conf.augnew	2018-06-04 11:57:14.499415677 +0200
@@ -1,2 +1,2 @@

-alias acpi:IPI000*:* ipmi_devintf
+alias acpi:IPI000*:* ipmi_si

Notice: /Stage[main]/Ipmitool::Kmod/Kmod::Alias[acpi:IPI000*:* - ipmi_si]/Kmod::Setting[kmod::alias acpi:IPI000*:* - ipmi_si]/Augeas[kmod::setting kmod::alias acpi:IPI000*:* - ipmi_si acpi:IPI000*:*]/returns: executed successfully
Notice: Augeas[kmod::setting kmod::alias acpi:IPI000*:* - ipmi_msghandler acpi:IPI000*:*](provider=augeas):
--- /etc/modprobe.d/ipmitool-modalias.conf	2018-06-04 11:57:14.520415537 +0200
+++ /etc/modprobe.d/ipmitool-modalias.conf.augnew	2018-06-04 11:57:14.529415477 +0200
@@ -1,2 +1,2 @@

-alias acpi:IPI000*:* ipmi_si
+alias acpi:IPI000*:* ipmi_msghandler

Notice: /Stage[main]/Ipmitool::Kmod/Kmod::Alias[acpi:IPI000*:* - ipmi_msghandler]/Kmod::Setting[kmod::alias acpi:IPI000*:* - ipmi_msghandler]/Augeas[kmod::setting kmod::alias acpi:IPI000*:* - ipmi_msghandler acpi:IPI000*:*]/returns: executed successfully
Notice: Applied catalog in 10.67 seconds
[11:57:16] root@bar.example.org ~ # grep alias /etc/modprobe.d/ipmitool-modalias.conf
alias acpi:IPI000*:* ipmi_msghandler
```
(The changes happen on every Puppet Run)

## Your Environment
* Puppet version (`puppet -V`): 5.5.1
* Operating System and version: `CentOS Linux release 7.4.1708 (Core)`

**Comment by baurmatt** [comment]
This seems to be broken for CentOS 6 as well, if the file (/etc/modprobe.d/ipmitool-modalias.conf) is empty. But the ipmitool rpm shipped with CentOS 6 ships /etc/modprobe.d/ipmitool-modalias.conf with all of those three aliases included, so we never ran into this problem.

**Comment by baurmatt** [comment]
TL;DR: kmod::(alias|setting) for the same (alias|module) name is broken due to this augeas line:

```
          "set ${category}[. = '${module}'] ${module}",
          "set ${category}[. = '${module}']/${option} ${value}",
```

This effectively blocks from having the same alias name defined more than once in the same file.

**Comment by raphink** [comment]
As you noted, the definition currently considers the `$module` as a primary field, thus modifying the value every time the same module is used.

If that behavior is changed, it won't be possible to change a module's value. What do you suggest the behavior and interface should be?

**Comment by baurmatt** [comment]
Sorry for the wait, got somehow lost in my backlog -.- I, as well, don't see an option to solve this with augeas. Perhaps switching over to concat would be a solution which solves this issue.

**Comment by raphink** [comment]
@baurmatt 	the problem is not implementation, it is the interface: `$module` is currently a primary key, yet you'd like to use it as a property instead.
